DRAFT: sdk v0.50 upgrade

niilptr opened the pull request.

koofree commented:
Hi all, I am a member of Firmachain. I just mention about a small peace.

```
        upgrade_test.go:91: 
            	Error Trace:	/home/runner/work/firmachain/firmachain/app/upgrades/v05/upgrade_test.go:91
            	            				/home/runner/work/firmachain/firmachain/app/upgrades/v05/upgrade_test.go:52
            	Error:      	Not equal: 
            	            	expected: "0.520000000000000000"
            	            	actual  : "0.500000000000000000"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1 +1 @@
            	            	-0.520000000000000000
            	            	+0.500000000000000000
            	Test:       	TestKeeperTestSuite/TestUpgrade
```

Check this one, plz

koofree commented:
First, I appreciate all these contributions.

I have a few questions to make sure I understand these updates.

1. Do you plan to use `ProposalCancelRatio` and `ExpeditedMinDeposit`? It is currently not usable. Or I don't have a sense of what they are.

2. `burnmodule` was deprecated. Isn't it not needed anymore?

koofree commented on app/upgrades/v05/upgrades.go:
```diff
@@ -0,0 +1,236 @@
+				panic(err)
+			}
+			logger.Info("proposals migration: added proposer field for given proposals.")
+		}
```
```suggestion
		} else {
		 // is it needs to raise error?
		}
```

niilptr replied:
Since the proposals migration is optional, and we have to specify the proposer address, we set these two branches depending on the current chain-id running. The kintsugi one has been used to test it on our devnet. It is not needed to set a default case here, since already by default, if the chains running the upgrade don't match the specified chain-ids no migration is performed. Raising an error here would lock the upgrade to work with the specified chains-id only, which would be annoying during testing / development.

koofree replied:
Nice! I make a sense of it.

koofree commented on app/upgrades/v05/upgrades.go:
```diff
@@ -0,0 +1,236 @@
+		newIcaHostParams := icahosttypes.Params{
+			HostEnabled: true,
+			// https://github.com/cosmos/ibc-go/blob/v4.2.0/docs/apps/interchain-accounts/parameters.md#allowmessages
+			AllowMessages: []string{"*"},
```
We don't need to set the Allow Messages directly not "*"

like below example
```suggestion
			AllowMessages: ["/cosmos.gov.v1.Vote"],
```

niilptr replied:
If you want just the cosmos.gov.v1.Vote message to be allowed, we can change it to this. Can I proceed to set AllowMessages to ["/cosmos.gov.v1.Vote"]?

koofree replied:
I wonder about the vulnerability when we set this as a `*`. If it doesn't, `*` is okay.
And, I don't exactly know what governance features require us to use messages when operating interchain. I think we could leave it as a `*` for now.

niilptr replied:
@koofree There should be no vulnerability in using `*` (at least in theory), since this is the configuration used by `gaia` and `juno` for example.
If you want to be more conservative, we could use the approach that is used by `osmosis`, which allows the following cosmos/ibc/wasm messages at the moment:
```
- /ibc.applications.transfer.v1.MsgTransfer

- /cosmos.bank.v1beta1.MsgSend

- /cosmos.staking.v1beta1.MsgDelegate
- /cosmos.staking.v1beta1.MsgBeginRedelegate
- /cosmos.staking.v1beta1.MsgCreateValidator
- /cosmos.staking.v1beta1.MsgEditValidator
- /cosmos.staking.v1beta1.MsgUndelegate

- /cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward
- /cosmos.distribution.v1beta1.MsgSetWithdrawAddress
- /cosmos.distribution.v1beta1.MsgWithdrawValidatorCommission
- /cosmos.distribution.v1beta1.MsgFundCommunityPool

- /cosmos.gov.v1beta1.MsgVote

- /cosmwasm.wasm.v1.MsgExecuteContract
- /cosmwasm.wasm.v1.MsgInstantiateContract

- /cosmos.authz.v1beta1.MsgGrant
- /cosmos.authz.v1beta1.MsgRevoke
```

koofree replied:
Hi @niilptr! I clearly understood. If we can choose the conservative way, I have no doubt doing that way. Can we go on your suggestion?

koofree commented on app/upgrades/v05/upgrades.go:
```diff
@@ -0,0 +1,236 @@
+		ctx := sdk.UnwrapSDKContext(c)
+		logger := ctx.Logger().With("upgrade", UpgradeName)
+
+		// === Params migration ===
```
Is there any way to validate that param migration is not missing existing params?

niilptr replied:
For the params migration verification, we tested it on our lolcanets and on our devnet to be sure that each param value persists after the migration. From our testing this migration is working properly.
In order to let you have a testable upgrade in the shortest time possible, we did not setup the upgrade test in the codebase to cover for all the parameters, since this check has been separately managed by us during our upgrade test in our localnets and devnet. Anyway, if you want to have a test confirmation, we can include this check inside the codebase' upgrade test for sure.

koofree replied:
I see, understood. Thank you for informing me about it.

niilptr commented:
Hi @koofree, thanks for the feedbacks! I will answer to your previous questions here:
1. [`TestUpgrade` failing] The test was failing because we updated the `MinInitialDepositRatio` param but did not change the expected value in the test case. We tested it in our localnet and it is working properly. I've update the check, thank you for this catch!
2. [`ProposalCancelRatio` and `ExpeditedMinDeposit`] The `ProposalCancelRatio` and `ExpeditedMinDeposit` params are new gov parameters that manage two new features available with the Cosmos SDK v0.50 (respectively, cancelling proposals, and expedited proposals). The current version of Firmachain, v04, is using the old Cosmos SDK v0.46, which has no support for these new features. Once the upgrade will be executed, users will have the possibility to start using the cancelling-proposals and expedited-proposals features. If you want to further tune the expedited-proposals behaviour we can also set custom values for the `ExpeditedThreshold` and `ExpeditedVotingPeriod` params in this upgrade (these, as any other param, can anyway be changed via governance proposals if you don't want to include their value now in this upgrade). You can find all the information for these new features in the [cancelling proposals documentation](https://docs.cosmos.network/v0.50/build/modules/gov#cancel-proposal) and the [expedited proposals documentation](https://docs.cosmos.network/v0.50/build/modules/gov#expedited-proposals).
3. [`x/burn` module] After rechecking the burn module, I think we made a mistake here. We removed it because it looked like an unused module. But after reviewing it again, its endblock functionality is indeed called by the end-blocker which burns all the burn-module address balances. If you still need this feature we can re-add it back for sure. Sorry for having overlooked this.

koofree commented:
@niilptr I appreciate all the answers about this PR. Through it, I highly understand this upgrade.
